### ISSUE TITLE
Add support for watching without pagination

### ISSUE THREAD

**Issue by shawkins** [fabric8io/kubernetes-client#5081]
### Is your enhancement related to a problem? Please describe

See https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md - and related changes in the go-client.  Kubernetes api-servers will soon support better initial watch semantics such that list operations will not be required.

### Describe the solution you'd like

Change the watch parameters and implement support for the annotated bookmark denoting the end of the initial list.

**Comment by csviri** [comment]
Hi @shawkins , do you plan to work on this? I might be able to pick this up, if not.

**Comment by shawkins** [comment]
> Hi @shawkins , do you plan to work on this? I might be able to pick this up, if not.

I don't have plans currently. It would definitely be good to get in there. The pagination logic definitely complicates things, so it eventually should go.

**Comment by csviri** [comment]
Just to make sure I understand what we want to achieve, so this should also replace the generic `list()` logic for resources. (Not just implement it for informer list watch)

**Comment by csviri** [comment]
Putting here the new blogpost just for the record: https://kubernetes.io/blog/2024/12/17/kube-apiserver-api-streaming/

**Comment by shawkins** [comment]
> Just to make sure I understand what we want to achieve, so this should also replace the generic list() logic for resources. (Not just implement it for informer list watch)

Are you thinking that would look starting a watch with a Watcher that collects values, and terminates after seeing sendInitialEvents action? That seems good. Or do you want to have something more async as well?

Regardless we'll need it, and the informer logic, to be driven either by server feature introspection or a config value.

**Comment by csviri** [comment]
> Are you thinking that would look starting a watch with a Watcher that collects values, and terminates after seeing sendInitialEvents action? That seems good. Or do you want to have something more async as well?

Just the first part now, with the old apis in place (no async)

**Comment by shawkins** [comment]
I'll have a pr with the changes to the informer shortly. Just wanted to see what are the thoughts on other usage of this feature. An example higher-level method could look like:

```
public static CompletableFuture<Void> streamingList(ListOptions listOptions, Consumer<T> onItem) {
    CompletableFuture<Void> future = new CompletableFuture<>();
    Watch watch = this.watch(listOptions, new Watcher<T>() {

      @Override
      public void eventReceived(Action action, T resource) {
        switch (action) {
        case ADDED:
          onItem.accept(resource);
          break;
        case BOOKMARK:
          future.complete(null);
          break;
        default:
          if (action == Action.ERROR && resource instanceof Status) {
            Status status = (Status) resource;
            future.completeExceptionally(new KubernetesClientException(status));
          } else {
            future.completeExceptionally(new KubernetesClientException(String.format("Unexpected action %s: %s", action, Serialization.asYaml(resource))));
          }
          break;
        }
      }

      @Override
      public void onClose(WatcherException cause) {
        future.completeExceptionally(cause);
      }
      
    });
    future.whenComplete((v, t) -> watch.close());
    return future;
  }
```

You then need to call it with the list options built in a particular way (or have the method always set them) - sendInitialEvents=true, resourceVersionMatch=NotOlderThan, allowWatchBookmarks=true and a resource version is optional:

```
streamingList(new ListOptionsBuilder()
        .withSendInitialEvents(true)
        .withResourceVersionMatch("NotOlderThan")
        .withAllowWatchBookmarks(true).build(), item -> {})...
```

I believe @csviri expressed interest in having something run automatically under the covers for normal Listable.list calls - that is also possible (for example if sendInitialEvents=true), but we do need to return a KubernetesResourceList and may take a bit of additional logic to always determine the approriate api kind for the list - or we could simply look for the list class and use "List" if it can't be found.

**Comment by manusa** [comment]
This looks good.

> You then need to call it with the list options built in a particular way

Regarding the _special_ ListOptions, a helper method or sth along those lines could be provided to return a defaulted `ListOptionsBuilder` that returned a builder with the initially required configuration set, for example: `new StreamingListOptionsBuilder()`.

I assume this is because not all clusters might support streaming.
Maybe we can also do the opposite, default to streaming, but allow for legacy support by providing the "specially crafted" options (while introducing a breaking behavioral change note).

**Comment by shawkins** [comment]
> I assume this is because not all clusters might support streaming.

Yes, it needs to be 1.27+ and on most of those versions the WatchList feature gate must be enabled: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

On 1.27+ when you don't have WatchList enabled, but submit a request with sendInitialEvents=true, then you see 422 error.

Let's say we add a Config option as mentioned on #7190, we could have something like this to signal the usage of streaming lists with the normal list api in BaseOperation:

```
  public CompletableFuture<L> submitList(ListOptions listOptions) {
    if (Boolean.TRUE.equals(getConfig().getStreamingList()) || Boolean.TRUE.equals(listOptions.getSendInitialEvents())) {
      // set or validate the required listOptions
      return streamingList(listOptions);
    }

    ...
```

The only drawback are the list type issue and you are not handling the results in a memory safe way. So I think we'll probably want to also expose something that looks like https://github.com/fabric8io/kubernetes-client/issues/5081#issuecomment-3124763623 either on the Listable or Watchable interfaces. Thinking about it more, it should actually return CompletableFuture\<String\> so that it contains the final bookmark resourceVersion.

**Comment by csviri** [comment]
Than you @shawkins for taking care of this, just some context information, I think this will be properly rolled out from k8s side just in version 1.34, see: https://ahmet.im/blog/kubernetes-list-performance/#watch-list
There are some nice links for KEPs
